### ISSUE TITLE
Loss wouldn't decrease when training on GPU while everything is OK on CPU

### ISSUE THREAD

**Issue by alexeyev** [changun/CollMetric#10]
Hello, 
thanks for open-sourcing your code and for it being so nicely written!

We've bumped into the following issue: everything works fine when we train the model on CPU (loss decreases very fast), however, when we run that very same code on GPU, the loss would stay approximately the same and the method wouldn't converge at all.

Batch size/learning rate modification didn't help. 

Checked the issue on tensorflow 1.4.x and 1.7.0 and on different CUDA versions.

Did you have the same problems with CML? Do you have any ideas why that happens and how to fix it?

Many thanks in advance.

**Comment by changun** [comment]
Hi Anton,

I do remember I encountered a similar issue, but can't recall how I fix
them.... Check the errors of each iteration and see if there is any NA? It
could be caused by GPU's low floating point precision.

**Comment by changun** [comment]
and which Optimizer do you use? I remember I did have problem with Tensorflow's AdaGrad optimize (no gradient at all), but Adam optimizer works perfectly for me.

**Comment by alexeyev** [comment]
Thanks for  such a quick response!

> Adam optimizer works perfectly for me

Yes, **replacing Adagrad with Adam (+ decrease in learning rate) solved the issue**! 
(learning rate reduction on its own doesn't help)

However, I still haven't managed to find out why this happens : )
